text/0021-any-and-all-operators.md

```diff
@@ -0,0 +1,165 @@
+  [0,1].all(
+    ...
+      [0,1].all(true)...))
+```
+
+Because nested quantifiers are prohibited, there is no need to support predicates with arbitrary parameter names. The `it` keyword sufficiently extends the grammar.
+
+### Extending the semantics
+
+The semantics of `all` and `any` are straightforward when there are no errors:
+- `E.all(P)` evaluates to true when `P[e/it]` is true for every element `e` in `E`.
+- `E.any(P)` evaluates to true when `P[e/it]` is true for some element `e` in `E`.
+
+But what happens if evaluating `P` errors on some element `e`? In this case, the evaluation of `all` or `any` terminates abruptly with a distinguished `QuantifierError`.
+
+The `QuantifierError` error may include static information (such as source location), but no dynamic information (such as the value that caused the error to be thrown). This error handling approach ensures that evaluating  `all` and `any` is _deterministic_: the result is the same regardless of iteration order.
+
+For example, consider the expression `[1, true].all(it like "foo")`. Regardless of the order in which the predicate is applied, the expression returns the same `QuantifierError`.
+
+This semantics lets us think about `all` and `any` in terms of their expansion to conjunctions and disjunctions. The quantified expression errors if and only if the expansion would error.
+
+For example:
+```
+// Throws QuantifierError
+[1, true].all(it like "foo")
+
+// Throws a TypeError: "like" can't be applied to an integer
+1 like "foo" && true like "foo"
+
+// Throws a TypeError: "like" can't be applied to an boolean
+true like "foo" && 1 like "foo"
+```
+
+The only difference is the error being thrown: `QuantifierError` for the quantified expression versus an element-specific error for the explicit expansion.
+
+### Extending the type system
+
+The type system needs to be extended with rules for type checking quantified expressions `E.all(P)` and `E.any(P)`. This should be fairly standard: we know the type of `it` from the type of the set `E`, and we can use this to typecheck the predicate `P`. We also need to propagate effects through predicates.
+
+For example, the following should typecheck when `context.limit` is an optional integer attribute, and `resource.things` is a set of entities with an optional integer attribute `.size`:
+
+```
+context has limit &&
+resource.things.all(
+  it has size && it.size < context.limit
+)
+```
+
+## Drawbacks
+
+- This proposal requires significant development effort. We'll need to modify the parser, the CST to AST conversion, CST/AST/EST, evaluator, validator, models, proofs, and DRT.
+- The `any` and `all` operators may encourage writing complex expressions with poor performance. For example, `context.portNumbers.any(it == 8010)` is $O(n)$ versus $O(1)$ for `context.portNumbers.contains(8010)`.
+- Supporting these operators will likely have a negative impact on analysis performance, especially for complex predicates.
+
+## Alternatives
+
```

Comment:
What about this alternative:

You evaluate the predicate on _every_ member of the set, and collect all of the errors to be returned in a deterministic order? So in your example `[1, true].all(it like "foo")` you'd return the equivalent of 
```
[(true,TypeError("Boolean found where String expected"),
 (1,TypeError("Long found where String expected"))]
```
The order in the list is based on a global order defined on Cedar values, e.g., that all Booleans come before all Longs.

Reply:
Alternatively, you could sort based on error type, and ignore the contents of the error message and the expression that accompanies it. I believe that matches the level of abstraction we have today when considering Cedar error messages in the model.

Reply:
I don't think that using a list is a viable alternative because it forces the specification to impose an (arbitrary) order on values and/or error types.  Then, every implementation has to enforce the same order.

If we want to go this route, how about returning a _set_ of error messages instead?  

Not that this a non-trivial change because it changes the return type of the evaluator from `Result<Value, Error>` to `Result<Value, Set<Error>>`. 

Another alternative is to package this error set into the QuantifierError.

Reply:
We can still preserve the mapping:  use a map from values to errors.  That still avoids having to impose an arbitrary ordering in the specification, and allows implementations to use a HashMap if desired.

Reply:
Thinking about this more:  returning a map from values to errors makes both evaluation and authorization asymptotically more expensive.  

In the worst case, the size of the answer returned by the authorizer will be linear in the size of both the policy slice _and_ the entity slice with this approach.  For example, imagine a set that contains 1,000 elements of the wrong type.  Then this alternative will return a map from 1,000 of these elements to the errors they caused. 

This alternative is fundamentally changing the size complexity of the evaluation and authorization results.  I don't think we should adopt it for this reason.  It may be impossible to return the full result over the wire if/when that is desired (e.g., AVP and similar services).

Reply:
The return value is now O(n) in the size of the policy, whereas before it was O(1), but the evaluation time is O(n) either way, is that right?

What if we had some sort of debug mode, where you return "quantifierError" in the general case, but return the map in debug mode?

Reply:
It's somewhat worse.  

For evaluation, the size of the return value becomes O(n) in the size of the input, where it used to be O(1) with respect to both input and policy size.  For authorization, the size of the return value becomes O(n * p + p) where n is the size of the input and p is the number of policies. 

Is the idea that we surface a debug mode in the language spec too?  Or don't surface / test / model / prove anything about it?

Reply:
Yes, I think we surface some sort of debug mode. My thinking goes like this:
1. In production, an authorization request returns a response that indicates a policy has produced an error
2. The customer application notes this fact in its own logs, along with information about the request, i.e., the PARC and provided entity data
3. Later, the dev team can replay the request against the identified policy in debug mode, which then produces the detailed error message indicating what element of the list actually failed

As long as step 1 returns sufficient information for customers to do steps 2 and 3, once a debug-mode exists, I'm OK with this RFC as is.

Reply:
Debug mode probably is another evaluator implementation which we have to prove/test is equivalent to the production one.

Reply:
This sounds reasonable to me.  

As an implementation note, the natural choice for the map from `foo` element to error would probably be a map from `Expr` to error. (Note that `Expr` implements `Hash` so it can be put in an unordered hashmap.)  But I'm a little less clear on what type the keys should be if you are calling `.any()` or `.all()` on a set that originates from the entity data or `Context`.  `Expr` is still one option (we could easily convert the `RestrictedExpr`s to `Expr` and return them in that form), but RFC 34 proposes not keeping around the `RestrictedExpr` by the time we get to evaluation.  At that point you'd have to have the key be `Value`, but those are not `Hash`.  `Value`s are totally ordered, so you can put them in a treemap, but then your concern above about appearing unordered is kind of a fiction.  Moreover, if `Value`s are indeed the key type, they'd have to be made public -- something which is also not currently the case but also proposed in RFC 34.

Reply:
Perhaps an acceptable workaround to the above would be to have the key be the string representation of the `Value`; the string representation would be trivially total-ordered, hashable, and useful for debugging, while dodging the issues about whether `Value` is made public.

Reply:
I think we need to keep the size of the evaluator output constant.  Letting it grow linearly with input size (policies + request +  entities) means that the authorizer's output is worst-case quadratic in input size.  This is undesirable even if rare.

I've revised the RFC to relax the specification of `QuantifierError` so that it can include more information  in the error message, as long as the additional information is computed deterministically and is bounded in size.